From 15d2138a45d3850830cca90e483a59fc43e87458 Mon Sep 17 00:00:00 2001
From: Phillip Lougher <phillip@squashfs.org.uk>
Date: Sat, 12 Jul 2014 02:47:15 +0100
Subject: [PATCH 010/120] actions: add dir_ent to action_data structure

dir_ent is needed for the contained() test operator and for the
new contained_followlink() test operator.

Note: action_data contains some fields which are copied from dir_ent,
and are thus now duplicated.  This will be cleaned up in the near
future.

Historical note: the action_data structure when it was originally
created back in 2012 didn't use the dir_ent structure because at
that time the dir_ent didn't exist when the actions code was called
to evaluate the exclude action().  Iff the exclude action returned
FALSE the dir_ent structure was created then.  Since then dir_scan1()
has been restructured to create the dir_ent structure before any
actions code is called.

Signed-off-by: Phillip Lougher <phillip@squashfs.org.uk>
---
 squashfs-tools/action.c     | 7 ++++++-
 squashfs-tools/action.h     | 4 +++-
 squashfs-tools/mksquashfs.c | 2 +-
 3 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/squashfs-tools/action.c b/squashfs-tools/action.c
index 84de265..d98a8f2 100644
--- a/squashfs-tools/action.c
+++ b/squashfs-tools/action.c
@@ -663,6 +663,7 @@ void eval_actions(struct dir_ent *dir_ent)
 	action_data.subpath = subpathname(dir_ent);
 	action_data.buf = &dir_ent->inode->buf;
 	action_data.depth = dir_ent->our_dir->depth;
+	action_data.dir_ent = dir_ent;
 
 	for (i = 0; i < other_count; i++) {
 		struct action *action = &other_spec[i];
@@ -692,6 +693,7 @@ void *eval_frag_actions(struct dir_ent *dir_ent)
 	action_data.subpath = subpathname(dir_ent);
 	action_data.buf = &dir_ent->inode->buf;
 	action_data.depth = dir_ent->our_dir->depth;
+	action_data.dir_ent = dir_ent;
 
 	for (i = 0; i < fragment_count; i++) {
 		match = eval_expr(fragment_spec[i].expr, &action_data);
@@ -736,7 +738,7 @@ int exclude_actions()
 
 
 int eval_exclude_actions(char *name, char *pathname, char *subpath,
-	struct stat *buf, int depth)
+	struct stat *buf, int depth, struct dir_ent *dir_ent)
 {
 	int i, match = 0;
 	struct action_data action_data;
@@ -746,6 +748,7 @@ int eval_exclude_actions(char *name, char *pathname, char *subpath,
 	action_data.subpath = subpath;
 	action_data.buf = buf;
 	action_data.depth = depth;
+	action_data.dir_ent = dir_ent;
 
 	for (i = 0; i < exclude_count && !match; i++)
 		match = eval_expr(exclude_spec[i].expr, &action_data);
@@ -1264,6 +1267,7 @@ int eval_empty_actions(struct dir_ent *dir_ent)
 	action_data.subpath = subpathname(dir_ent);
 	action_data.buf = &dir_ent->inode->buf;
 	action_data.depth = dir_ent->our_dir->depth;
+	action_data.dir_ent = dir_ent;
 
 	for (i = 0; i < empty_count && !match; i++) {
 		data = empty_spec[i].data;
@@ -1503,6 +1507,7 @@ void eval_move_actions(struct dir_info *root, struct dir_ent *dir_ent)
 	action_data.subpath = subpathname(dir_ent);
 	action_data.buf = &dir_ent->inode->buf;
 	action_data.depth = dir_ent->our_dir->depth;
+	action_data.dir_ent = dir_ent;
 
 	/*
 	 * Evaluate each move action against the current file.  For any
diff --git a/squashfs-tools/action.h b/squashfs-tools/action.h
index 795fe34..d5394a9 100644
--- a/squashfs-tools/action.h
+++ b/squashfs-tools/action.h
@@ -198,6 +198,7 @@ struct action_data {
 	char *pathname;
 	char *subpath;
 	struct stat *buf;
+	struct dir_ent *dir_ent;
 };
 
 
@@ -279,7 +280,8 @@ extern int parse_action(char *);
 extern void dump_actions();
 extern void *eval_frag_actions(struct dir_ent *);
 extern void *get_frag_action(void *);
-extern int eval_exclude_actions(char *, char *, char *, struct stat *, int);
+extern int eval_exclude_actions(char *, char *, char *, struct stat *, int,
+							struct dir_ent *);
 extern void eval_actions(struct dir_ent *);
 extern int eval_empty_actions(struct dir_ent *dir_ent);
 extern void eval_move_actions(struct dir_info *, struct dir_ent *);
diff --git a/squashfs-tools/mksquashfs.c b/squashfs-tools/mksquashfs.c
index 35fa51d..9a662ee 100644
--- a/squashfs-tools/mksquashfs.c
+++ b/squashfs-tools/mksquashfs.c
@@ -3396,7 +3396,7 @@ struct dir_info *dir_scan1(char *filename, char *subpath,
 			subpath = subpathname(dir_ent);
 			
 			if(eval_exclude_actions(dir_name, filename, subpath,
-								&buf, depth)) {
+							&buf, depth, dir_ent)) {
 				add_excluded(dir);
 				free_dir_entry(dir_ent);
 				continue;
-- 
2.8.3

